src/recommendation/src/main/scala/RankingAdapter.scala

```diff
@@ -0,0 +1,382 @@
+  }
+
+}
+
+object RankingAdapter extends ComplexParamsReadable[RankingAdapter]
+
+/**
+  * Model from train validation split.
+  *
+  * @param uid Id.
+  */
+class RankingAdapterModel private[ml](val uid: String)
+  extends Model[RankingAdapterModel] with ComplexParamsWritable with Wrappable with RankingFunctions {
+
+  def recommendForAllUsers(i: Int): DataFrame = getRecommenderModel.asInstanceOf[ALSModel].recommendForAllUsers(3)
+
+  def recommendForAllItems(i: Int): DataFrame = getRecommenderModel.asInstanceOf[ALSModel].recommendForAllItems(3)
+
+  def this() = this(Identifiable.randomUID("RankingAdapterModel"))
+
+  val recommenderModel = new TransformerParam(this, "recommenderModel", "recommenderModel", { x: Transformer => true })
+
+  def setRecommenderModel(m: Model[_]): this.type = set(recommenderModel, m)
+
+  def getRecommenderModel: Model[_] = $(recommenderModel).asInstanceOf[Model[_]]
+
+  val mode: Param[String] = new Param(this, "mode", "recommendation mode")
+
+  /** @group getParam */
+  def getMode: String = $(mode)
+
+  /** @group setParam */
+  def setMode(value: String): this.type = set(mode, value)
+
+  val nItems: IntParam = new IntParam(this, "nItems", "recommendation mode")
+
+  /** @group getParam */
+  def getNItems: Int = $(nItems)
+
+  /** @group setParam */
+  def setNItems(value: Int): this.type = set(nItems, value)
+
+  val nUsers: IntParam = new IntParam(this, "nUsers", "recommendation mode")
+
+  /** @group getParam */
+  def getNUsers: Int = $(nUsers)
+
+  /** @group setParam */
+  def setNUsers(value: Int): this.type = set(nUsers, value)
+
+  def transform(dataset: Dataset[_]): DataFrame = {
+    transformSchema(dataset.schema)
+    val model = getRecommenderModel
+    getMode match {
+      case "allUsers" => {
+        val recs = this.recommendForAllUsers(getNItems)
+        prepareTestData(getUserCol, getItemCol, dataset.toDF(), recs, 10).toDF()
+      }
+      case "allItems" => {
+        val recs = model.asInstanceOf[ALSModel].recommendForAllItems(getNUsers)
+        prepareTestData(getItemCol, getUserCol, dataset.toDF(), recs, 10).toDF()
+      }
+      case "normal"   => {
+        val recs = SparkHelper.flatten(model.transform(dataset), 10, getItemCol, getUserCol)
+        prepareTestData(getItemCol, getUserCol, dataset.toDF(), recs, 10).toDF()
+      }
+    }
+  }
+
+  def transformSchema(schema: StructType): StructType = {
+    val model = getRecommenderModel
+    getMode match {
+      case "allUsers" => new StructType()
+        .add("userCol", IntegerType)
+        .add("recommendations", ArrayType(
+          new StructType().add("itemCol", IntegerType).add("rating", FloatType))
+        )
+      case "allItems" => new StructType()
+        .add("itemCol", IntegerType)
+        .add("recommendations", ArrayType(
+          new StructType().add("userCol", IntegerType).add("rating", FloatType))
+        )
+      case "normal"   => new StructType()
+        .add("userCol", IntegerType)
```

Comment:
hmm, there are a lot of hard-coded string here, I think some of these should be constants, others params?